libkpod/attach.go

```diff
@@ -0,0 +1,182 @@
+	}
+
+	resize := make(chan remotecommand.TerminalSize)
+
+	AttachContainerSocket(ctr, resize, noStdIn, detachKeys)
+	close(resize)
+	c.ContainerStateToDisk(ctr)
+
+	return nil
+}
+
+// getStandardStreams returns the standard IO readers/writers
+func getStandardStreams() (stdIn, stdOut, stdErr *os.File) {
+	return os.Stdin, os.Stdout, os.Stderr
+}
+
+func closeInputStream(stdIn io.ReadCloser) {
+	stdIn.Close()
+}
+
+// AttachContainerSocket connects to the crio socket and deals with the IO
+func AttachContainerSocket(ctr *oci.Container, resize <-chan remotecommand.TerminalSize, noStdIn bool, detachKeys []byte) error {
+	inputStream, outputStream, errorStream := getStandardStreams()
+	defer closeInputStream(inputStream)
+
+	tty, err := strconv.ParseBool(ctr.State().Annotations["io.kubernetes.cri-o.TTY"])
+	if err != nil {
+		return errors.Errorf("unable to parse annotations in %", ctr.ID)
+	}
+	if !tty {
+		return errors.Errorf("no tty available", ctr.ID())
+	}
+
+	oldTermState, err := term.SaveState(inputStream.Fd())
+	defer term.RestoreTerminal(inputStream.Fd(), oldTermState)
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to save terminal state")
+	}
+
+	// Put both input and output into raw
+	if !noStdIn {
+		term.SetRawTerminal(inputStream.Fd())
+	}
+
+	controlPath := filepath.Join(ctr.BundlePath(), "ctl")
+	controlFile, err := os.OpenFile(controlPath, unix.O_WRONLY, 0)
+	if err != nil {
+		return fmt.Errorf("failed to open container ctl file: %v", err)
+	}
+
+	kubecontainer.HandleResizing(resize, func(size remotecommand.TerminalSize) {
+		logrus.Infof("Got a resize event: %+v", size)
+		_, err := fmt.Fprintf(controlFile, "%d %d %d\n", 1, size.Height, size.Width)
+		if err != nil {
+			logrus.Infof("Failed to write to control file to resize terminal: %v", err)
+		}
+	})
+	attachSocketPath := filepath.Join("/var/run/crio", ctr.ID(), "attach")
+	conn, err := net.DialUnix("unixpacket", nil, &net.UnixAddr{Name: attachSocketPath, Net: "unixpacket"})
+	if err != nil {
+		return fmt.Errorf("failed to connect to container %s attach socket: %v", ctr.ID(), err)
+	}
+	defer conn.Close()
+
+	receiveStdout := make(chan error)
+	if outputStream != nil || errorStream != nil {
+		go func() {
+			receiveStdout <- redirectResponseToOutputStreams(outputStream, errorStream, conn)
+		}()
+	}
+
+	stdinDone := make(chan error)
+	go func() {
+		var err error
+		if inputStream != nil && !noStdIn {
+			_, err = utils.CopyDetachable(conn, inputStream, detachKeys)
+			conn.CloseWrite()
+		}
+		stdinDone <- err
+	}()
+
+	select {
+	case err := <-receiveStdout:
+		return err
+	case err := <-stdinDone:
+		if _, ok := err.(utils.DetachError); ok {
+			return nil
+		}
+		if outputStream != nil || errorStream != nil {
+			return <-receiveStdout
+		}
+	}
+
+	return nil
+}
+
+func redirectResponseToOutputStreams(outputStream, errorStream io.Writer, conn io.Reader) error {
+	var err error
+	buf := make([]byte, 8192+1) /* Sync with conmon STDIO_BUF_SIZE */
```

Comment:
It kind of bothers me that we have conmon constants duplicated here, but I don't think it's worth doing a CGo import for only 4 constants and no actual functionality.

Reply:
agree